File: prime_sieve/sieve.ts
function sieveOfEratosthenes(n: number): number[] {
    if (n < 2) {
        return [];
    }
    
    // Create boolean array
    const primes: boolean[] = new Array(n + 1).fill(true);
    primes[0] = primes[1] = false;
    
    // Sieve algorithm
    for (let i = 2; i * i <= n; i++) {
        if (primes[i]) {
            for (let j = i * i; j <= n; j += i) {
                primes[j] = false;
            }
        }
    }
    
    // Collect prime numbers
    const result: number[] = [];
    for (let i = 2; i <= n; i++) {
        if (primes[i]) {
            result.push(i);
        }
    }
    
    return result;
}

function main(): void {
    const limit = 10000000;  // 10 million
    const primes = sieveOfEratosthenes(limit);
    
    console.log(`Found ${primes.length} primes up to ${limit}`);
    console.log(`First 10 primes: ${primes.slice(0, 10).join(' ')}`);
    console.log(`Last 10 primes: ${primes.slice(-10).join(' ')}`);
}

main();

File: prime_sieve/sieve.go
package main

import (
	"fmt"
)

func sieveOfEratosthenes(n int) []int {
	if n < 2 {
		return []int{}
	}

	// Create boolean slice
	primes := make([]bool, n+1)
	for i := 2; i <= n; i++ {
		primes[i] = true
	}

	// Sieve algorithm
	for i := 2; i*i <= n; i++ {
		if primes[i] {
			for j := i * i; j <= n; j += i {
				primes[j] = false
			}
		}
	}

	// Collect prime numbers
	var result []int
	for i := 2; i <= n; i++ {
		if primes[i] {
			result = append(result, i)
		}
	}

	return result
}

func main() {
	limit := 10000000 // 10 million
	primes := sieveOfEratosthenes(limit)

	fmt.Printf("Found %d primes up to %d\n", len(primes), limit)
	fmt.Print("First 10 primes: ")
	for i := 0; i < 10 && i < len(primes); i++ {
		fmt.Printf("%d ", primes[i])
	}
	fmt.Println()

	fmt.Print("Last 10 primes: ")
	for i := len(primes) - 10; i < len(primes); i++ {
		fmt.Printf("%d ", primes[i])
	}
	fmt.Println()
}
